File: web/src/types/analysis.rs
//! Versioned decision data, separate from board rendering and training replay.
use serde::{Deserialize, Serialize};
use super::GameStateResponse;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PositionKey {
    pub session_id: String,
    pub revision: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckpointIdentity {
    pub checkpoint_id: String,
    pub training_step: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ActionReference {
    Discrete { index: u32 },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionAssessment {
    pub action: ActionReference,
    pub network_prior: Option<f32>,
    pub search_prior: Option<f32>,
    pub visit_share: Option<f32>,
    pub selection_probability: Option<f32>,
    pub visits: Option<u32>,
    pub q_value: Option<f32>,
    pub expanded: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ValueQuantity { ExpectedOutcome, DiscountedReturn }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValueEstimate {
    pub value: f32,
    pub perspective_agent: u32,
    pub quantity: ValueQuantity,
    pub bounds: Option<[f32; 2]>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchEffort {
    pub completed_simulations: u32,
    pub root_visits: u32,
    pub neural_evaluations: u32,
    pub total_time_us: u64,
    pub temperature: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DecisionSource { AlphaZeroMcts, Random, Human, DqnQValues }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DecisionAnalysis {
    pub schema_version: u32,
    pub analysis_id: String,
    pub position: PositionKey,
    pub env_id: String,
    pub env_contract_version: u32,
    pub algorithm_id: String,
    pub actor: u32,
    pub source: DecisionSource,
    pub checkpoint: Option<CheckpointIdentity>,
    pub selected_action: ActionReference,
    pub network_value: Option<ValueEstimate>,
    pub search_value: Option<ValueEstimate>,
    pub actions: Vec<ActionAssessment>,
    pub search: Option<SearchEffort>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PositionRecord {
    pub state: GameStateResponse,
    /// Decision made FROM this snapshot; the next record is its result.
    pub decision: Option<DecisionAnalysis>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryResponse {
    pub schema_version: u32,
    pub session_id: String,
    pub first_available_revision: u64,
    pub current_revision: u64,
    pub records: Vec<PositionRecord>,
    pub next_revision: Option<u64>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn non_board_agent_zero_q_values_need_no_board_or_search() {
        let value = ValueEstimate { value: 2.5, perspective_agent: 0,
            quantity: ValueQuantity::DiscountedReturn, bounds: None };
        let data = serde_json::to_value(value).unwrap();
        assert_eq!(data["perspective_agent"], 0);
        assert_eq!(data["quantity"], "discounted_return");
        assert_eq!(data["value"], 2.5);
        assert!(data["bounds"].is_null());
    }
}

File: web/frontend/src/lib/analysis.ts
// Board-independent inspector contract. Probabilities are optional: Q learners
// and human decisions do not fabricate policy/search data.
import type { GameState } from './api';

export type ActionTarget =
  | { kind: 'cell'; index: number }
  | { kind: 'column'; index: number }
  | { kind: 'edge'; from: number; to: number }
  | { kind: 'named'; name: string };

export interface ActionPresentation { action: number; label: string; target: ActionTarget | null }
export interface PositionKey { session_id: string; revision: number }
export interface ActionAssessment {
  action: { kind: 'discrete'; index: number };
  network_prior: number | null;
  search_prior: number | null;
  visit_share: number | null;
  selection_probability: number | null;
  visits: number | null;
  q_value: number | null;
  expanded: boolean | null;
}
export interface ValueEstimate {
  value: number;
  perspective_agent: number;
  quantity: 'expected_outcome' | 'discounted_return';
  bounds: [number, number] | null;
}
export interface DecisionAnalysis {
  schema_version: number;
  analysis_id: string;
  position: PositionKey;
  env_id: string;
  env_contract_version: number;
  algorithm_id: string;
  actor: number;
  source: 'alpha_zero_mcts' | 'random' | 'human' | 'dqn_q_values';
  checkpoint: { checkpoint_id: string; training_step: number | null } | null;
  selected_action: { kind: 'discrete'; index: number };
  network_value: ValueEstimate | null;
  search_value: ValueEstimate | null;
  actions: ActionAssessment[];
  search: {
    completed_simulations: number; root_visits: number; neural_evaluations: number;
    total_time_us: number; temperature: number;
  } | null;
}
export interface PositionRecord { state: GameState; decision: DecisionAnalysis | null }
export interface HistoryResponse {
  schema_version: number;
  session_id: string;
  first_available_revision: number;
  current_revision: number;
  records: PositionRecord[];
  next_revision: number | null;
}
export type ProbabilityMetric = 'visit_share' | 'network_prior' | 'selection_probability';

export function samePosition(left: PositionKey, right: PositionKey): boolean {
  return left.session_id === right.session_id && left.revision === right.revision;
}

export function percent(value: number | null | undefined): string {
  return value == null || !Number.isFinite(value) ? '—' : `${(value * 100).toFixed(1)}%`;
}

export function sortedActions(actions: ActionAssessment[], metric: ProbabilityMetric): ActionAssessment[] {
  return [...actions].sort((a, b) =>
    (b[metric] ?? -Infinity) - (a[metric] ?? -Infinity) || a.action.index - b.action.index);
}

/** Aggregate an explicitly declared source/cell/column without renormalizing. */
export function targetMass(presentations: ActionPresentation[], actions: ActionAssessment[],
  metric: ProbabilityMetric, kind: 'cell' | 'column' | 'source', index: number): number | null {
  const ids = new Set(presentations.filter(({ target }) => target && (
    kind === 'source' ? target.kind === 'edge' && target.from === index :
    target.kind === kind && target.index === index
  )).map(a => a.action));
  const values = actions.filter(a => ids.has(a.action.index)).map(a => a[metric]);
  const available = values.filter((v): v is number => v != null && Number.isFinite(v));
  return available.length ? available.reduce((sum, value) => sum + value, 0) : null;
}

export function boundAnalysis(record: PositionRecord | null): DecisionAnalysis | null {
  if (!record?.decision || record.decision.schema_version !== 1 ||
      !samePosition(record.state, record.decision.position)) return null;
  return record.decision;
}
